### ISSUE TITLE
canonical stem: not being returned? not being used in matching process? fuzzy-match not working?

### ISSUE THREAD

**Issue by abubelinha** [gnames/gnverifier#80]
I looked at the url you provided in [#78](https://github.com/gnames/gnverifier/issues/78#issuecomment-1032900155) about different canonical forms.

Regarding canonical stem usage for matching names, I see gnparser returns its value, but not gnverifier. 

I think gnverifier is not even using it for matching, which is a shame because we loose some interesting matches.

For example, _Isoetes longissima_ Bory ([~22700 Google returns](https://www.google.com/search?q=Isoetes+longissima)) has been many times cited as _Isoetes longissimum_ Bory ([~4000 Google returns](https://www.google.com/search?q=Isoetes+longissimum)).

[gnverifier returns](http://verifier.globalnames.org/?all_sources=on&ds=197&ds=198&format=json&names=Isoetes%20longissima%20Bory%0D%0AIsoetes%20longissimum%20Bory&all_matches=true) 13 exact matches for the 1st one, and 1 exact match for the 2nd one.
I would expect it to return **also** 1 fuzzy match for the 1st one, and 13 fuzzy matches for the 2nd one (all of them are "canonicalStem-exact").

So that if I look for _Isoetes longissimum_ shouldn't I got 13 returns from the 13 sources which include _Isoetes longissima_? And the other way round, to get 1 source coming from [EUNIS datasource](http://verifier.globalnames.org/data_sources/158):
https://eunis.eea.europa.eu/species/316085

I am interested in this canonical matching process for the metodology of a regional checklist creation (related bibliographic non-technical question in [gnparser #44](https://github.com/gnames/gnparser/issues/44#issuecomment-1025105256))

**Comment by dimus** [comment]
@abubelinha 
Match is using stemmed canonical actually, so it does match gender mistakes in species.

However your example according to the algorithm I use (see docs to gnparser) does not generate the same stemmed canonical

http://parser.globalnames.org/?format=html&names=Isoetes+longissimum+Bory%0D%0AIsoetes+longissima+Bory

As soon as a match happaned, verification stops to save CPU cycles. You can use `Advanced search` to get more variants, for
example 

https://verifier.globalnames.org/?all_matches=&all_sources=&capitalize=on&format=html&names=n%3AIsoetes+longiss.+Bory

**Comment by abubelinha** [comment]
Thanks @dimus ... but I am afraid we are misunderstanding each other. 

Why do you say "_these names do not generate the same stemmed canonical_"?
From the 1st link you posted, I see the same result for both names: 
http://parser.globalnames.org/?format=html&names=Isoetes+longissimum+Bory%0D%0AIsoetes+longissima+Bory
```json
    "stemmed": "Isoetes longissim"
```
And when you say "_However if suffix is different, it is considered Fuzzy match, even if stem is the same_" ... isn't this the same thing I said I expected above?
- stem is the same: "Isoetes longissim"
- suffix is different: "-um" vs "-a"
- hence, this _**should**_ be considered a Fuzzy match by gnverifier.  

As I said above: gnverifier _**should have returned**_ 13 fuzzy matches for "[Isoetes longissimum](https://verifier.globalnames.org/?all_sources=on&format=json&names=Isoetes%20longissimum%20Bory&all_matches=true)" (because there are 13 datasources containing the name "Isoetes longissima"). And 1 fuzzy match searching the other way round, for "[Isoetes longissima](https://verifier.globalnames.org/?all_sources=on&format=json&names=Isoetes%20longissima%20Bory&all_matches=true)".

So, why is this not happening? 
It only returns Exact matches (I don't see any Fuzzy matches in those json results)

**Comment by dimus** [comment]
oups my bad, yes, the same stem

Here is how stems match works at the moment

1. make a stem canonical out of input name
2. compare it with stems in the database
3. check simple canonical forms for identity
4. if there is an exact match, return that, discard others

So, if `Isoetes longissimus` is entered, 2 simple canonical forms are returned: `Isoetes longissimus` (exact) and `Isoetes longissuma` (fuzzy). Exact is given and fuzzy is discarded.

So this fuzzy match is actually cheap, because the actual match was 'exact match' by stemmed canonical.

**Comment by abubelinha** [comment]
> So, if Isoetes longissimus is entered, 2 simple canonical forms are returned: Isoetes longissimus (exact) and Isoetes longissuma (fuzzy). Exact is returned and fuzzy is discarded.

OK ... that makes sense when an Exact match is found, like it happens with the bestResult in this case.
But what about the rest of the 13 datasources found? When **no** Exact matches are found, a Fuzzy match should be returned.

To put it more clear: I want to discard bestResult and match my list only against WFO, WCVP and LCVP (196, 197, 198).
They all contain entries for _Isoetes longissima_. But my list contains only _Isoetes longissimum_.
So when I send my list to gnverifier, this name is not having any "preferred results" matches (neither Exact -of course- nor Fuzzy). 
In other words: the array is not returned in both [v0](http://verifier.globalnames.org/api/v0/verifications/Isoetes%20longissimum%20Bory?pref_sources=197|196|198&all_matches=true) (`results`) and [v1](http://verifier.globalnames.org/api/v1/verifications/Isoetes%20longissimum%20Bory?pref_sources=197|196|198&all_matches=true) (`preferredResults`). 

This makes no sense to me, because it's pretty obvious that it's the same name which exists in the 3 preferred sources
(I guess these datasources' editors didn't even consider necessary to citate as synonyms those names sharing the same stem and authorship: probably they were just considered as ortographic variations of the accepted name).

**Comment by dimus** [comment]
Added https://github.com/gnames/gnmatcher/issues/45 about returning matches with different suffixes. I think this conversation can be closed now.

**Comment by abubelinha** [comment]
OK thanks a lot. 
Just to clarify my mind about changing this issue to gnmatcher.
Is this more or less correct?
- **gnverifier** relies on **gnmatcher** for doing matching, so this is a gnmatcher issue
- **resolver** gives a different output because it does its own matching process (which is somehow different to gnmatcher's process)

Sorry about my confussion ... all these process are so distributed among related projects, that it's hard for me to tell where should I open each issue.

**Comment by dimus** [comment]
There are 3 steps in name verification
1. gnmatcher makes exact and fuzzy matching but without any information about data-sources. This allows to avoid calls to database which significantly speeds things up.
2. gnames receives canonical form and match types from gnmatcher finds relevant data-sources, sorts results and finds the 'best' one.
3. gnverifier displays the data, if it was not a direct API call

The reason I did start this new approach is a necessity to verify non-trivial number of names. For example for indexing BHL there are 20 mil name-strings to verify, for HathiTrust library it is 300 million name-strings.

**Comment by dimus** [comment]
I do not exect users to know exactly where to place an Issue (gnverifier, gnmatcher, or gnames), so I think you are placing your verification issues at exactly right place. I can figure out which component is involved and create a "child" issue, like it happened with this one..
